addons/sale_margin/models/sale_order.py

```diff
@@ -24,6 +24,8 @@ def _compute_purchase_price(self):
                 continue
             line = line.with_company(line.company_id)
             product_cost = line.product_id.standard_price
+            if line.purchase_price:
+                product_cost = line.purchase_price
```

Comment:
> we want the recomputation when the product/currency/uom is modified.

AFAIK, if I change the product now, the cost won't be recomputed...

Honestly, after more thoughts, I don't know if there is a way to fix this problem (we would have to not trigger the recomputation for non timesheet lines if the computation trigger is the timesheet dependencies), but if there is, it must be done in sale_timesheet_margin, the base behavior is the expected one, you shouldn't have to modify it...